packages/astro/src/vite-plugin-config-alias/index.ts

```diff
@@ -0,0 +1,99 @@
+import * as tsr from 'tsconfig-resolver';
+import * as path from 'path';
+import * as url from 'url';
+
+import type * as vite from 'vite';
+
+/** Result of successfully parsed tsconfig.json or jsconfig.json. */
+export declare interface Alias {
+  find: RegExp;
+  replacement: string;
+}
+
+/** Returns a path with its slashes replaced with posix slashes. */
+const normalize = (pathname: string) => String(pathname).split(path.sep).join(path.posix.sep);
+
+/** Returns the results of a config file if it exists, otherwise null. */
+const getExistingConfig = (searchName: string, cwd: string | undefined): tsr.TsConfigResultSuccess | null => {
+  const config = tsr.tsconfigResolverSync({ cwd, searchName });
+
+  return config.exists ? config : null;
+};
+
+/** Returns a list of compiled aliases. */
+const getConfigAlias = (cwd: string | undefined): Alias[] | null => {
+  /** Closest tsconfig.json or jsconfig.json */
+  const config = getExistingConfig('tsconfig.json', cwd) || getExistingConfig('jsconfig.json', cwd);
+
+  // if no config was found, return null
+  if (!config) return null;
+
+  /** Compiler options from tsconfig.json or jsconfig.json. */
+  const compilerOptions = Object(config.config.compilerOptions);
+
+  // if no compilerOptions.baseUrl was defined, return null
+  if (!compilerOptions.baseUrl) return null;
+
+  // resolve the base url from the configuration file directory
+  const baseUrl = path.posix.resolve(path.posix.dirname(normalize(config.path)), normalize(compilerOptions.baseUrl));
+
+  /** List of compiled alias expressions. */
+  const aliases: Alias[] = [];
+
+  // compile any alias expressions and push them to the list
+  for (let [alias, values] of Object.entries(Object(compilerOptions.paths) as { [key: string]: string[] })) {
+    values = [].concat(values as never);
+
+    /** Regular Expression used to match a given path. */
+    const find = new RegExp(`^${[...alias].map((segment) => (segment === '*' ? '(.+)' : segment.replace(/[\\^$*+?.()|[\]{}]/, '\\$&'))).join('')}$`);
+
+    /** Internal index used to calculate the matching id in a replacement. */
+    let matchId = 0;
+
+    for (let value of values) {
+      /** String used to replace a matched path. */
+      const replacement = [...path.posix.resolve(baseUrl, value)].map((segment) => (segment === '*' ? `$${++matchId}` : segment === '$' ? '$$' : segment)).join('');
+
+      aliases.push({ find, replacement });
+    }
+  }
+
+  // compile the baseUrl expression and push it to the list
+  aliases.push({
+    find: /^(?!\.*\/)(.+)$/,
+    replacement: `${baseUrl
+      .split('')
+      .map((segment) => (segment === '$' ? '$$' : segment))
+      .join('')}/$1`,
+  });
```

Comment:
I had to look this up because I wasn't sure it was correct, but it is!

I would possibly add some more context to the comment above, and possibly include this link: https://www.typescriptlang.org/tsconfig#baseUrl

The `baseUrl` option changes the way non-relative specifiers are resolved, so if `baseUrl` exists, all non-relative specifiers will be resolved relative to the `baseUrl`.